File: src/p2815.rs
#[test]
fn test() {
    assert_eq!(max_sum(vec![51, 71, 17, 24, 42]), 88);
    assert_eq!(max_sum(vec![1, 2, 3, 4]), -1);
}

// hashmap: {'max_num', [num1, num2, ...]}
// 先找数字的数位上最大的那个数字，加入哈希表
// 排序，取最后两位
pub fn max_sum(nums: Vec<i32>) -> i32 {
    use std::collections::HashMap;
    let mut map: HashMap<char, Vec<i32>> = HashMap::new();

    // 条件1：这两个数组的数位上的最大数字相等
    for num in nums {
        // 获取各数位上的最大数字，加入哈希表
        if let Some(max_num) = num.to_string().chars().max() {
            map.entry(max_num)
                .and_modify(|v| v.push(num))
                .or_insert(vec![num]);
        }
    }

    // 条件2：这两个数字的和最大
    let mut ans: i32 = -1;
    for v in map.values_mut() {
        // 数组包含2个及以上的元素
        if v.len() >= 2 {
            v.sort();
            let last2_sum: i32 = *&v[v.len() - 2..].iter().sum();
            ans = ans.max(last2_sum);
        }
        // 数组元素不够
        else {
            continue;
        }
    }

    ans
}

File: src/p2815.py
# https://leetcode.cn/problems/max-pair-sum-in-an-array/description/

import unittest
from math import inf
from typing import List


class Solution:
    def maxSum(self, nums: List[int]) -> int:
        ans: int = -1
        # max_val定义了一个可以容纳10个元素的数组，其索引表示一个元素的最大数位
        # max_val维护了最大数位为i的最大元素
        max_val: List[int] = [-inf] * 10

        for i in nums:
            max_d: int = max(map(int, str(i)))  # 当前元素的最大数位
            ans = max(ans, i + max_val[max_d])  # 更新答案，此时已经保证了最大数位相等
            max_val[max_d] = max(max_val[max_d], i)  # 更新最大元素

        return ans


class Test(unittest.TestCase):
    def test(self) -> None:
        solution = Solution()
        self.assertEqual(solution.maxSum([51, 71, 17, 24, 42]), 88)
        self.assertEqual(solution.maxSum([1, 2, 3, 4]), -1)


if __name__ == "__main__":
    unittest.main()
